src/skelet_cps.rs

```diff
@@ -0,0 +1,376 @@
+            self.to.push(to);
+        }
+    }
+}
+
+type Adjacents = HashMap<(Orientation, ITape), Adjacent>;
+type Configurations = IndexSet<Configuration>;
+
+/// ClosedPositionSet
+pub struct CPS {
+    adjs: Adjacents,
+    confs: Configurations,
+    segment_size: usize,
+    step_limit: usize,
+    interner: Interner,
+
+    #[cfg(feature = "ui_tui")]
+    conf2end: Vec<String>,
+}
+
+impl CPS {
+    fn add_conf(confs: &mut Configurations, mut conf: Configuration, tape: ITape, _interner: &Interner) {
+        conf.tape[conf.head.orient()] = tape;
+        let (_idx, _new) = confs.insert_full(conf);
+        ui_dbg!(
+            "\t\t\tconf {} {}",
+            if _new { "added" } else { "exists" },
+            confs.get_index(_idx).unwrap().fmt(_interner)
+        );
+    }
+
+    pub fn run(&mut self, machine: &Machine) -> Result<(), Err> {
+        let CPS {
+            adjs,
+            confs,
+            segment_size,
+            step_limit,
+            interner,
+            #[cfg(feature = "ui_tui")]
+            conf2end,
+        } = self;
+        let (segment_size, step_limit) = (*segment_size, *step_limit);
+
+        let mut conf_id = 0;
+        while let Some(old_conf) = confs.get_index(conf_id) {
+            ui_dbg!("running conf id={conf_id} {}", old_conf.fmt(interner));
+            // `<S ABC` -> `<S AB` + `C`
+            let (conf, c, _last_conf) = old_conf.run(machine, step_limit, segment_size, interner)?;
+            let orient = conf.head.orient;
+            #[cfg(feature = "ui_tui")]
+            conf2end.push(_last_conf);
+
+            // from -> to; accumulates new edges due to dependencies
+            let mut try_add: Vec<(ITape, ITape)> = Vec::new();
+            // `ab s> c` -> `<S ABC` => copy edges starting from c into C
+            ui_dbg!("\tnew adj from replace:");
+            let old_c = old_conf.tape[conf.head.op_orient()];
+            if old_c != c {
+                let old = Adjacent::get(adjs, 1 - orient, old_c);
+                ui_dbg!(
+                    "\t\tadding dep {} to ({}, {})",
+                    c.fmt(1 - orient, interner),
+                    1 - orient,
+                    old_c.fmt(1 - orient, interner),
+                );
+                old.deps.push(c);
+
+                let tos = old.to.clone();
+                if !tos.is_empty() {
+                    let adj = Adjacent::get(adjs, 1 - orient, c);
+                    tos.into_iter().for_each(|to| adj.add_adj(confs, to, &mut try_add, 1 - orient, c, interner));
+                }
+            }
+            ui_dbg!("\tnew adj from shift:");
+            // `<S ABC` => add B -> C
+            let b = conf.tape[conf.head.op_orient()];
+            Adjacent::get(adjs, 1 - orient, b).add_adj(confs, c, &mut try_add, 1 - orient, b, interner);
+            ui_dbg!("\tnew adj from dependencies:");
+            while let Some((from, to)) = try_add.pop() {
```

Comment:
No clue what is happening here? Some sort of parallelism race condition??

Reply:
just a construct to process all items from collection. new possible edges (due of dependencies) are accumulated into `try_add`. the problem is that i can add new elements to it as i process it, so i can't use classic iteration.

Reply:
Ooooh, yes, I understand. I have something similar, but for me, I make a copy before going through the "todos" then repeat until there are none left.